src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -610,71 +643,126 @@ def get_installation_id(self):
             self._storage[_INSTALLATION_ID] = installation_id
         return installation_id
 
+
+class MsiAccountTypes(object):
+    # pylint: disable=no-method-argument,no-self-argument
+    system_assigned = 'MSI'
+    user_assigned_client_id = 'MSIClient'
+    user_assigned_object_id = 'MSIObject'
+    user_assigned_resource_id = 'MSIResource'
+
+    @staticmethod
+    def valid_msi_account_types():
+        return [MsiAccountTypes.system_assigned, MsiAccountTypes.user_assigned_client_id,
+                MsiAccountTypes.user_assigned_object_id, MsiAccountTypes.user_assigned_resource_id]
+
     @staticmethod
-    def get_msi_token(resource, port, identity_id=None, for_login=False):
+    def msi_auth_factory(cli_account_name, identity, resource):
+        if cli_account_name == MsiAccountTypes.system_assigned:
+            return MSIImdsAuthentication(resource=resource)
+        elif cli_account_name == MsiAccountTypes.user_assigned_client_id:
+            return MSIImdsAuthentication(resource=resource, client_id=identity)
+        elif cli_account_name == MsiAccountTypes.user_assigned_object_id:
+            return MSIImdsAuthentication(resource=resource, object_id=identity)
+        elif cli_account_name == MsiAccountTypes.user_assigned_resource_id:
+            return MSIImdsAuthentication(resource=resource, ms_res_id=identity)
+        else:
+            raise ValueError("unrecognized msi account name '{}'".format(cli_account_name))
+
+# pylint: disable=too-many-lines
+
+
+class MSIImdsAuthentication(BasicTokenAuthentication):
```

Comment:
@lmazuel, please review. I will incorporate this class into https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/azure_active_directory.py#L648-L685, the effort should be very trivial. 
I propose we do one thing at a time in the sequence of 
    1. Get this PR good and merged
    2. I will move the code & related test to msrest
    3. You publish a new msrest
    4. I submit a new PR in CLI to switching